### PR TITLE
Add Place Picker widget for the 'Create Event Form' #706

### LINKED ISSUE #706: Add Place Picker widget for the 'Create Event Form'
**Is it a ?**

- [ ] Bug Report

- [x] Feature Request
    
- [ ] Chore
       
### Feature Request

**Feature Description**

Manually copying and pasting, or typing can be a bit tedious. User experience can be improved here. 
For the 'Create Event Form', options for Place Picker. Option for location detection using network service or GPS can also be added.
   
**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by mishuvs**
Fixes #706 

Changes: 
1. `Latitude` and `Longitude` text fields removed.
2. Place Picker button added in place.
3. `Location Name` and `Searchable Location name` are hidden initially, to make the user use the map. Once location from the map is selected, the fields are made visible to the user to confirm the auto-completed `Location Name` and `Searchable Location Name`. (For searchable location, selects most accurate address substring with no numbers)

Gif of the change: 
picker

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -51,6 +51,9 @@
             android:name=".module.organizer.detail.OrganizerDetailActivity"
             android:label="@string/title_activity_organizer_detail"
             android:theme="@style/AppTheme.Main.Light" />
+        <meta-data
+            android:name="com.google.android.geo.API_KEY"
+            android:value="INSERT_YOUR_API_KEY"/>
```
Comment (mishuvs) [inline app/src/main/AndroidManifest.xml]: @iamareebjamal Please suggest how to hide the API key.

Comment (iamareebjamal) [reply app/src/main/AndroidManifest.xml]: Load it through the environment variables

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java [hunk 2]
```diff
@@ -97,9 +99,9 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                 try {
                     startActivityForResult(builder.build(getActivity()), PLACE_PICKER_REQUEST);
                 } catch (GooglePlayServicesRepairableException e) {
-                    e.printStackTrace();
+                    Log.e(LOG_TAG,"GooglePlayServicesRepairable",e);
                 } catch (GooglePlayServicesNotAvailableException e) {
-                    e.printStackTrace();
+                    Log.e(LOG_TAG,"GooglePlayServices NotAvailable => Updating or Unauthentic");
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java]: Use Timber

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java [hunk 3]
```diff
@@ -243,10 +243,9 @@ private void showLocationLayouts() {
     }
 
     private String getSearchableLocationName(String address) {
-        String primary = address.substring(0, address.indexOf(","));
-        String secondary = address.substring(address.indexOf(",") + 2, address.indexOf(",", address.indexOf(",") + 1));
+        String primary = address.substring(0, address.indexOf(','));
         if (primary.matches(".*\\d+.*")) { //contains number => not likely to be searchable
-            return secondary;
+            return address.substring(address.indexOf(',') + 2, address.indexOf(",", address.indexOf(',') + 1));
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java]: What is this function doing?

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -182,6 +182,7 @@
     <string name="create_faq">Create FAQ</string>
     <string name="question">Question</string>
     <string name="answer">Answer</string>
+    <string name="pick_from_map">Pick from map</string>
```
Comment (iamareebjamal) [inline app/src/main/res/values/strings.xml]: 'M' capital

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java [hunk 5]
```diff
@@ -77,6 +91,30 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+                //SERVICE_MISSING, SERVICE_VERSION_UPDATE_REQUIRED, SERVICE_DISABLED
+                GoogleApiAvailability.getInstance().getErrorDialog(
+                    getActivity(), errorCode, PLACE_PICKER_REQUEST, (dialog) -> {
+                        showLocationLayouts();
+                    });
+            } else {
+                //SERVICE_UPDATING, SERVICE_INVALID - can't use place picker - must enter manually
+                showLocationLayouts();
+            }
+        });
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java]: Instance of GoogleApiAvailability should be stored in a global variable

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java [hunk 6]
```diff
@@ -77,6 +91,30 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+            if (errorCode == ConnectionResult.SUCCESS) {
+                //SUCCESS
+                PlacePicker.IntentBuilder builder = new PlacePicker.IntentBuilder();
+                try {
+                    startActivityForResult(builder.build(getActivity()), PLACE_PICKER_REQUEST);
+                } catch (GooglePlayServicesRepairableException e) {
+                    Log.e(LOG_TAG, "GooglePlayServicesRepairable", e);
+                } catch (GooglePlayServicesNotAvailableException e) {
+                    Log.e(LOG_TAG, "GooglePlayServices NotAvailable => Updating or Unauthentic");
+                }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java]: Use Timber

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java (hunk 5 above)
Comment (mishuvs) [reply app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java]: I have made the specified changes. Please review.

File: app/config.gradle [hunk 7]
```diff
@@ -25,4 +25,5 @@ ext {
     leak_canary_version = '1.5.4'
     support_lib_version = '27.0.2'
     play_services_vision='11.8.0'
+    play_services_places='11.8.0'
```
Comment (iamareebjamal) [inline app/config.gradle]: Duplicate version

File: app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java [hunk 8]
```diff
@@ -178,5 +216,36 @@ public void setDefaultTimeZone(int index) {
...
+        binding.form.layoutSearchableLocation.setVisibility(View.VISIBLE);
+        binding.form.layoutLocationName.setVisibility(View.VISIBLE);
+    }
+
+    private String getSearchableLocationName(String address) {
+        String primary = address.substring(0, address.indexOf(','));
+        if (primary.matches(".*\\d+.*")) { //contains number => not likely to be searchable
+            return address.substring(address.indexOf(',') + 2, address.indexOf(",", address.indexOf(',') + 1));
+        } else return primary;
+    }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java]: Should be in presenter so that it is testable

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/app/module/event/create/CreateEventFragment.java]: Also write a comment showing what the function is doing

File: app/build.gradle [hunk 9]
```diff
@@ -27,7 +27,13 @@ android {
         versionName "1.0.3alpha"
         testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
         vectorDrawables.useSupportLibrary = true
-        manifestPlaceholders = [appName: app_name]
+        each { type ->
+            type.buildConfigField "String", "GOOGLE_PLACES_API_KEY", "\"" + GOOGLE_PLACES_API_KEY + "\""
```
Comment (iamareebjamal) [inline app/build.gradle]: Use envoronment variables please, or else we won't have any way of injecting the API key at build time

Comment (mishuvs) [reply app/build.gradle]: sorry, I have corrected

**Comment by mishuvs** [comment]
@iamareebjamal Please review.

**Comment by iamareebjamal** [comment]
Pending Tasks:
- [ ] Add API Key for Places API in build environment

**Comment by iamareebjamal** [comment]
@mishuvs A request. Can you please change the PR so that if we don't have an API key, it reverts to the old method. This adds a needed flexibility because some people won't want to set up an API key but still have the app not crash while testing it

**Comment by mishuvs** [comment]
Reverting back would need re-adding the location fields. Devs don't input these while creating an event, as it's not mandatory. Can we just hide the location layout in absence of an API key?
I think the location input is not being tested. How is it crashing?

**Comment by iamareebjamal** [comment]
The app is generic for all developers and not just for eventyay. They may choose not to register for Places API. Then the app will crash if they run without API key. Also, it will crash for each dev who hasn't registered for it as well

**Comment by mishuvs** [comment]
Can you drop a hint on how to check if there's an API key at runtime? 
(The app crashes only when the PlacePicker activity is invoked, so can't do it by handling exception.)

**Comment by iamareebjamal** [comment]
https://stackoverflow.com/questions/38687159/what-is-metadata-and-what-is-the-use-of-it-in-android

**Pull request merged**